### PR TITLE
[loader] Check for ReferenceAssemblyAttribute when loading

### PR DESCRIPTION
**Pull request by lambdageek**
Only load assemblies with the ReferenceAssemblyAttribute for reflection, not for execution.

Added a few tests:
 - using reflection api to open for execution throws an exception
 - using reflection api to open for reflection-only is okay
 - same thing as above two, but using `Load(byte[])`
 - compiled assembly references a reference assembly

### CODE REVIEW

File: mono/metadata/image.h [hunk 1]
```diff
@@ -15,7 +15,8 @@ typedef enum {
 	MONO_IMAGE_OK,
 	MONO_IMAGE_ERROR_ERRNO,
 	MONO_IMAGE_MISSING_ASSEMBLYREF,
-	MONO_IMAGE_IMAGE_INVALID
+	MONO_IMAGE_IMAGE_INVALID,
+	MONO_IMAGE_RUNNING_REFERENCE_ASSEMBLY
```
Comment (lambdageek) [inline mono/metadata/image.h]: Is this ok?  Do I need to bump a version number somewhere?

**Comment by akoeplinger** [comment]
>  something with corlib facades? (Actually we should also mark the corlib faces with the ReferenceAssemblyAttribute)

No, that'd be wrong. Facades can and do have code that should be executed.

**Comment by lambdageek** [comment]
@akoeplinger Maybe I don't mean "Facade", what are the ones that are in the -api directories?

**Comment by marek-safar** [comment]
@lambdageek we call them reference assemblies as they have ReferenceAssemblyAttribute. You can use them to test your patch (we have mscorlib.dll as reference assembly as well)

File: mono/metadata/appdomain.c [hunk 2]
```diff
@@ -1961,7 +1961,11 @@ ves_icall_System_Reflection_Assembly_LoadFrom (MonoString *fname, MonoBoolean re
 
 		if (status == MONO_IMAGE_IMAGE_INVALID)
 			exc = mono_get_exception_bad_image_format2 (NULL, fname);
-		else
+		else if (status == MONO_IMAGE_RUNNING_REFERENCE_ASSEMBLY) {
+			char *msg = g_strdup_printf ("Could not load file or assembly %s or one of its dependencies. Reference assemblies should not be loaded for execution.  They can only be loaded in the Reflection-only loader context.", ass->aname.name);
```
Comment (akoeplinger) [inline mono/metadata/appdomain.c]: I'd change the first part to `Didn't load file or assembly %s or one of its dependencies because it is a reference assembly. ...` so it's clear that this is intentional.

File: mono/metadata/assembly.c [hunk 3]
```diff
@@ -1907,12 +1942,26 @@ mono_assembly_load_from_full (MonoImage *image, const char*fname,
...
+		}
+	}
+
 	/* 
 	 * The load hooks might take locks so we can't call them while holding the
 	 * assemblies lock.
 	 */
 	if (ass->aname.name) {
-		ass2 = mono_assembly_invoke_search_hook_internal (&ass->aname, NULL, refonly, FALSE);
+		ass2 = mono_assembly_invoke_search_hook_internal (&ass->aname, NULL, ass->ref_only, FALSE);
```
Comment (kumpera) [inline mono/metadata/assembly.c]: why this change?

File: mono/metadata/image.h (hunk 1 above)
Comment (kumpera) [inline mono/metadata/image.h]: No need to bump any version number.

**Comment by lambdageek** [comment]
I think I'm going to have to re-work this a little.

1. The behavior of this PR is that if there are multiple paths to try for loading an assembly, then it will skip over any reference assemblies it finds in any of the paths until it finds a non-reference assembly.  (ie: if `MONO_PATH` is "d1:d2"  and `d1\a.dll` is a reference assembly and `d2\a.dll` is not, it will skip over `d1\a.dll` and then succeed and load `d2\a.dll`).
2. The behavior on .NET is: if there are multiple directories in the `<probing>` element in the app config, then the first directory that contains a matching assembly wins: so if the config has `<probing privatePath="d1;d2" />`, then if `d1\a.dll` is a reference assembly, you get a BadImageFormatException.   If the path was `d2;d1`, then you load the assembly.

So I'm going to change the patch to match .NET - check for the `ReferenceAssemblyAttribute` higher in the call tree after it finds a candidate image to work with and don't try to keep searching for a non-reference assembly in other dirs.

**Comment by lambdageek** [comment]
Everything is passing.  Ok to merge?
